### PR TITLE
i#5843 scheduler: Add input stream ordinal access

### PR DESCRIPTION
**Pull request by derekbruening**
Adds to the scheduler interface a query to obtain the current input stream's memtrace_stream_t handle.

Adds a new scheduler flag SCHEDULER_USE_INPUT_ORDINALS and sets it by default for parallel mode so the output stream's ordinals are suppressed and instead the current input stream's ordinals are presented on the output stream.  This fixes a problem where the parallel analysis tool framework saw accumulated ordinals across inputs.

Adds a similar flag SCHEDULER_USE_SINGLE_INPUT_ORDINALS which causes the first flag to be set if there is a single input and single output. This solves a serial mode problem where an analysis tool does want to see input gaps when there is no interleaving as there is only one input.

Adds a test.

Also manually tested a real analysis tool to confirm by tweaking the view tool to operate in parallel:

  Before:
    ===========================================================================
    [analyzer] Worker 0 starting on trace shard 0 stream is 0x562a2b0ff480
               1           0:     3443916 <marker: version 4>
               2           0:     3443916 <marker: filetype 0x240>
             ...
            1479         585:     3443916 <thread 3443916 exited>
    [analyzer] Worker 0 starting on trace shard 1 stream is 0x562a2b0ff480
    ------------------------------------------------------------
            1480         585:     3443921 <marker: version 4>
            1481         585:     3443921 <marker: filetype 0x240>
    ===========================================================================

  After:
    ===========================================================================
    [analyzer] Worker 0 starting on trace shard 0 stream is 0x555cebc44480
               1           0:     3443916 <marker: version 4>
               2           0:     3443916 <marker: filetype 0x240>
             ...
            1479         585:     3443916 <thread 3443916 exited>
    [analyzer] Worker 0 starting on trace shard 1 stream is 0x555cebc44480
    ------------------------------------------------------------
               1           0:     3443921 <marker: version 4>
               2           0:     3443921 <marker: filetype 0x240>
    ===========================================================================

Issue: #5843

### CODE REVIEW

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/scheduler/scheduler.h [hunk 1]
```diff
@@ -333,6 +333,18 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
...
         SCHEDULER_SPECULATE_NOPS = 0x2,
+        /**
+         * Causes the get_record_ordinal() and get_instruction_ordinal() results
+         * for an output stream to equal those values for the current input stream
+         * for that output, rather than accumulating across inputs.
+         */
+        SCHEDULER_USE_INPUT_ORDINALS = 0x4,
+        /**
+         * If there is just one input and just one output stream, this sets
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::SCHEDULER_USE_INPUT_ORDINALS;
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:344]: Added a comment:

        // This was added for the analyzer view tool on a single trace specified via
        // a directory where the analyzer isn't listing the dir so it doesn't know
        // whether to request SCHEDULER_USE_INPUT_ORDINALS.

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 2]
```diff
@@ -670,6 +678,16 @@ scheduler_tmpl_t<RecordType, ReaderType>::is_record_synthetic(int output_ordinal
     return inputs_[index].reader->is_record_synthetic();
 }
 
+template <typename RecordType, typename ReaderType>
+memtrace_stream_t *
+scheduler_tmpl_t<RecordType, ReaderType>::get_input_stream(int output_ordinal)
+{
+    int index = outputs_[output_ordinal].cur_input;
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:685]: Done.

File: clients/drcachesim/scheduler/scheduler.h [hunk 3]
```diff
@@ -455,21 +480,48 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
 
         /**
          * Returns the count of #memref_t records from the start of the trace to this
-         * point. This includes records skipped over and not presented to any tool. It
-         * does not include synthetic records (see is_record_synthetic()).
+         * point. It does not include synthetic records (see is_record_synthetic()).
+         *
+         * If
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::SCHEDULER_USE_INPUT_ORDINALS
+         * is set, then this value matches the record ordinal for the current input (and
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:487]: Done, on both places.

File: clients/drcachesim/scheduler/scheduler.h [hunk 4]
```diff
@@ -455,21 +480,48 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
...
+                        scheduler_->options_.flags))
+                return scheduler_->get_input_stream(ordinal_)->get_record_ordinal();
             return cur_ref_count_;
         }
         /**
          * Returns the count of instructions from the start of the trace to this point.
-         * This includes instructions skipped over and not presented to any tool.
+         * If
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::SCHEDULER_USE_INPUT_ORDINALS
+         * is set, then this value matches the record ordinal for the current input (and
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:509]: Done.

File: clients/drcachesim/scheduler/scheduler.h [hunk 5]
```diff
@@ -751,6 +812,9 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
     bool
     is_record_synthetic(int output_ordinal);
 
+    memtrace_stream_t *
+    get_input_stream(int output_ordinal);
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:816]: This is for the analyzer framework where the tool writer has no way to access the scheduler.  Maybe the analyzer could wrap the stream in its own implementation to provide this but we did have another possible use case where a simulator writer said they might like this feature too so it seems best to add to the scheduler.

File: clients/drcachesim/scheduler/scheduler.h [hunk 6]
```diff
@@ -751,6 +812,9 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
     bool
     is_record_synthetic(int output_ordinal);
 
+    memtrace_stream_t *
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:815]: Done.

File: clients/drcachesim/scheduler/scheduler.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:816]: There are already 4 other functions, 3 of them public, using the pattern "get_input_*", so it seems best to match those.

File: clients/drcachesim/scheduler/scheduler.h [hunk 7]
```diff
@@ -611,6 +663,15 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
         return static_cast<int>(inputs_.size());
     }
 
+    /** Returns the #memtrace_stream_t interface for the 'ordinal'-th input stream. */
+    virtual memtrace_stream_t *
+    get_input_stream_interface(int ordinal)
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:668]: Done.

**Pull request merged**